### PR TITLE
Programming Fundamentals Assignment 1

### PR DESCRIPTION
**Pull request by meta-akshita-dixit**
Sir,this is the assignment number 1 having 6 questions. Kindly review it and let me know if any changes are to me made.
Thank you.

### CODE REVIEW

File: PF-session1/assignment1/BinToOctal.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	}
+   /**
+    * this method finally generates octal number using toDecimal function.
+    * @param n
+    * @return
+    */
+  int convertBinaryToOctal(int n)               
+	{
+		int octal=0,i=0;
+		while(n!=0)
```
Comment (metacube-shivang) [inline PF-session1/assignment1/BinToOctal.java]: What if n is negative?

File: PF-session1/assignment1/TestBinToOct.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @author Akshita Dixit
+ */
+
+package assignment1;
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+
+public class TestBinToOct {
+	BinToOctal test =new BinToOctal();
```
Comment (metacube-shivang) [inline PF-session1/assignment1/TestBinToOct.java]: Different methods in a unit test should run independently. For this, normally, we do not instantiate global variables in a unit test class.

File: PF-session1/assignment1/BinToOctal.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+package assignment1;
+
+public class BinToOctal {
+	
+	/**
+	 * this method calculates the decimal value of triplets.
+	 * @param binary
+	 * @return
+	 */
+   int toDecimal(int binary)
```
Comment (metacube-shivang) [inline PF-session1/assignment1/BinToOctal.java]: The methods do not rely on any class variable. So, why not make them static?

File: PF-session1/assignment1/BinToOctal.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	/**
+	 * this method calculates the decimal value of triplets.
+	 * @param binary
+	 * @return
+	 */
+   int toDecimal(int binary)                      
+	{
+		int decimal=0,i=0;
+		while(binary!=0)
+		{
```
Comment (metacube-shivang) [inline PF-session1/assignment1/BinToOctal.java]: Use recommended syntax for where to open and close braces.

File: PF-session1/assignment1/TestBinToOct.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	@Test
+	public void test() {
```
Comment (metacube-shivang) [inline PF-session1/assignment1/TestBinToOct.java]: Test methods should be named such that it gives you an idea of what it is testing.  For example: this method could be "testConversionSuccess()" to indicate that it is testing for a positive case.

File: PF-session1/assignment1/TestBinToOct.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		assertEquals(65,test.convertBinaryToOctal(110101));
+	}
+	@Test
+	public void test2() {                                            //false test case
+		assertEquals(1,test.convertBinaryToOctal(100001));
```
Comment (metacube-shivang) [inline PF-session1/assignment1/TestBinToOct.java]: This test will fail. The test cases always need to pass, even when they are testing negative test cases. In this case, you should use "assertNotEquals". That way it tests that the returned value is not 1 and also the unit test passes.

File: PF-session1/assignment2/RemovingDuplicates.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+public class RemovingDuplicates 
+{
+	boolean checkDuplicate(int input[] , int i)
+	{
+		boolean flag = true;
+		for(int loopCounter = 0; loopCounter < i ; loopCounter++)
+		{
+			if(input[i] == input[loopCounter])           //compares with each element of the array.
+			{
+				flag = false;
```
Comment (metacube-shivang) [inline PF-session1/assignment2/RemovingDuplicates.java:24]: If a duplicate is detected, no need to check the remainder of the array; simply return false the first time the duplicate is detected.

File: PF-session1/assignment2/RemovingDuplicates.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+		while(i < size)
+		{
+			if(checkDuplicate(input,i))                 //passes array and each element's index to check if it is a duplicate element.
+			{
+				i++;
+			}
+			else
+			{
+				for(int loopCounter = i ; loopCounter < size-1; loopCounter++)
+					input[loopCounter] = input[loopCounter + 1];
```
Comment (metacube-shivang) [inline PF-session1/assignment2/RemovingDuplicates.java]: We are shifting all of the elements to the left, one element at a time. Consider a case where the input array is a large array consisting of only one unique element. Say, an array with a thousand 1s.

In this case, you'll be unnecessarily shifting elements repeatedly, even though ultimately the resulting array only needs to contain a single 1.

Think about an alternate way to handle this, without copying over duplicate elements.

File: PF-session1/assignment2/TestRemovingDuplicate.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package assignment2;
+
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+public class TestRemovingDuplicate {
```
Comment (metacube-shivang) [inline PF-session1/assignment2/TestRemovingDuplicate.java:7]: Follow other suggestions from assignment 1 regarding naming, test coverage, syntax, etc.

File: PF-session1/assignment3/LongestSequence.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+					break;
+			}
+			if(counter > count)                                 //to find the longest of all the possible sequences.
+			{
+				position=pos;
+				count=counter;
+			}
+			counter=1;
+		}
+		return Arrays.copyOfRange(input, position, position + count);
```
Comment (metacube-shivang) [inline PF-session1/assignment3/LongestSequence.java:40]: What if array is empty?

File: PF-session1/assignment3/TestLongestSequence.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import static org.junit.Assert.*;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestLongestSequence 
+{
+	LongestSequence test;
+	int arr[],expectedArr[];
+	@Before
```
Comment (metacube-shivang) [inline PF-session1/assignment3/TestLongestSequence.java]: @Before is useful when setting up common data between multiple unit tests. Here, there's only one unit test. So why do we need @Before?

Also, note that even if there are multiple unit tests, use of @Before will fail in this case because unit tests could potentially be run in multi-threaded environment, which will cause problems in our test because we are instantiating common data structures (arr[] and expectedArr[]) in the "before" method.

File: PF-session1/assignment3/TestLongestSequence.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+		arr=new int[]{1,2,3,2,3,4,5,3,4,2,2,3,4,5,6,7,8,1,2,4,5,6,7,8,9};
+		test=new LongestSequence();
+		expectedArr=new int[]{1,2,4,5,6,7,8,9};
+	}
+	@Test
+	public void test() {
+		assertArrayEquals(expectedArr,test.longestSequence(arr));
+	}
+
+}
```
Comment (metacube-shivang) [inline PF-session1/assignment3/TestLongestSequence.java]: Add more test cases (empty array, array with single element, array with negative values, with all elements equal, etc.)

File: PF-session1/assignment4/FirstComeFirstServe.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+		{
+			if(outputArr[loopCounter-1][4] >= outputArr[loopCounter][1])                                 //if arrival time of next job is less than finishing time of previous job.
+			{
+				outputArr[loopCounter][3] = outputArr[loopCounter - 1][4] + 1;                           //start time=finishing time of previous job+1
+				outputArr[loopCounter][2] = outputArr[loopCounter][3] - outputArr[loopCounter][1];       //waiting time=start time-arrival time
+				outputArr[loopCounter][4] = outputArr[loopCounter][3] + job_size[loopCounter] - 1;       //finishing time=start time+job size-1
+			}
+			else
+			{
+				outputArr[loopCounter][2] = 0;                                                           //waiting time
```
Comment (metacube-shivang) [inline PF-session1/assignment4/FirstComeFirstServe.java:35]: Setting of outputArr[][2] outputArr[][3] and outputArr[][4] is being done repeatedly at three places. This code should be extracted to a common method, with just required values being passed to it.

File: PF-session1/assignment5/AscOrDesc.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+		{
+			int position = 0;                          //it points the current index. 
+			int type;                                  //it is used to tell that if the array is unsorted or is in asc or desc  order.(0=unsorted,1=asc,2=desc)
+			if(array[position] <= array[position + 1])
+			{
+				type=1;
+				for(position=1; position < array.length - 1; position++)
+				{
+					if(array[position] > array[position + 1])
+						type=0;
```
Comment (metacube-shivang) [inline PF-session1/assignment5/AscOrDesc.java]: if type = 0 is detected, we can return immediately; no need to check for remaining elements.

File: PF-session1/assignment5/AscOrDesc.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+				}
+				return type;
+			}
+			else 
+			{
+				type=2;
+				for(position = 1; position < array.length - 1; position++)
+				{
+					if(array[position] <= array[position +1 ])
```
Comment (metacube-shivang) [inline PF-session1/assignment5/AscOrDesc.java]: why are we comparing with "<=" instead of just "<"?

File: PF-session1/assignment5/TestAscOrDesc.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+		checkSort=new AscOrDesc();
+		arr1=new int[]{8,7,6,5,4,-3,2,1};
+	}
+
+	@Test
+	public void test2() {               //it fails as it is unsorted and here is being checked for descending order.
+		assertEquals(2,checkSort.ascOrDesc(arr1));
+	}
+	
+	@Before
```
Comment (metacube-shivang) [inline PF-session1/assignment5/TestAscOrDesc.java]: Each @Before runs before each @Test, and there's no guarantee which will run in which order. What do you think the behavior will be when there are two @Before in the same test class? :)

**Review by metacube-shivang — CHANGES_REQUESTED** [review]
Hi @meta-akshita-dixit,

I've added some review comments. There are some that are problem specific, and some that are generic.

The generic ones have only been added to problem 1, but should be implemented in all the problems.

If there are any questions, we can discuss them tomorrow.

File: PF-session1/assignment3/TestLongestSequence.java [hunk 17]
```diff
@@ -7,19 +7,54 @@
...
+	@Test
+	public void testSuccessNegativeElements() {
+		LongestSequence test;
+		int arr[],expectedArr[];
+		arr=new int[]{-1,-3,-2,0,1};
+		test=new LongestSequence();
+		expectedArr=new int[]{-3,-2,0,1};
+		assertArrayEquals(expectedArr,test.longestSequence(arr));
+	}
 }
```
Comment (metacube-shivang) [inline PF-session1/assignment3/TestLongestSequence.java]: Much better test cases! 👍

**Review by metacube-shivang — APPROVED** [review]
